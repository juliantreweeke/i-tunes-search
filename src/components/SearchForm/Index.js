import PropTypes from "prop-types";
import React from "react";
import Button from "../Button/Index";
import styles from "./searchbar.module.css";
import useForm from "../../hooks/useForm";
import {isObjectEmpty} from  "../../helpers/helpers";
import EN from "../../EN.json";

const SearchForm = ({ loading, placeholder, handleSearch }) => {
  const fieldName = "searchQuery";
  const initialFieldValues = { searchQuery: ''}

  const {inputs, handleInputChange, handleKeyPress, handleSubmit} = useForm(handleSearch, initialFieldValues);
  const isButtonDisabled = loading || isObjectEmpty(inputs);

  return (
    <form
      className={styles.searchbar}
      onSubmit={handleSubmit}
    >
      <label htmlFor={fieldName}>
        <span className="visually-hidden">
          Enter search criteria here
        </span>
      </label>
      <div className={styles.input_container}>
        <input
          className={styles.input}
          type="text"
          id={fieldName}
          placeholder={placeholder || EN.DEFAULT_SEARCH_PLACEHOLDER}
          name={fieldName}
          onChange={handleInputChange}
          value={inputs[initialFieldValues]}
          onKeyPress={handleKeyPress}
        />
      </div>
      <Button disabled={isButtonDisabled} type="submit">{EN.SEARCH}</Button>
    </form>
  );
};

SearchForm.propTypes = {
  handleSearch: PropTypes.func.isRequired,
  placeholder: PropTypes.string
};

export default SearchForm;
