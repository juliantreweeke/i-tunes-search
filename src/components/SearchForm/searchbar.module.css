.searchbar {
  display: grid;
  grid-template-columns:0 1fr auto;
  padding-top: var(--space-xs);
}

.searchbar .input_container {
  padding-right: var(--component-padding);
}

.searchbar .input {
  border-radius: var(--radius);
  padding: var(--space-xs) var(--component-padding);
  border: 1px solid var(--color-border);
  background-color: var(--color-input);;
  box-sizing: border-box;
  color: var(--color-text-dark);
  width:100%;
  height:100%;
  box-sizing: border-box; 
  font-size: var(--text-sm);
  min-width: 200px;
}
