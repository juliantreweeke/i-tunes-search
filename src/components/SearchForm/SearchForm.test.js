import React from "react";
import { render, fireEvent } from "@testing-library/react";
import SearchForm from "./Index";
import { axe, toHaveNoViolations } from "jest-axe";

const mockHandleSearch = jest.fn();
const mockPlaceHolder = "placeholder";

describe("SearchForm", () => {
  it("should render", () => {
    const { container } = render(
      <SearchForm
        handleSearch={mockHandleSearch}
        placeholder={mockPlaceHolder}
      />
    );
    expect(container).toBeTruthy();
  });

  it("matches snapshot with placeholder prop", () => {
    const { container } = render(
      <SearchForm
        handleSearch={mockHandleSearch}
        placeholder={mockPlaceHolder}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot without placeholder prop", () => {
    const { container } = render(
      <SearchForm handleSearch={mockHandleSearch} />
    );
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with loading prop", () => {
    const { container } = render(
      <SearchForm handleSearch={mockHandleSearch} loading={true} />
    );
    expect(container).toMatchSnapshot();
  });

  it("Component should have no accessibility violations", async () => {
    expect.extend(toHaveNoViolations);
    const { container } = render(
      <SearchForm
        handleSearch={mockHandleSearch}
        placeholder={mockPlaceHolder}
      />
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it("input should allow values to be entered", async () => {
    const setup = () => {
      const utils = render(<SearchForm placeholder={mockPlaceHolder} handleSearch={mockHandleSearch}/>)
      const input = utils.getByPlaceholderText(mockPlaceHolder);
      return {
        input,
        ...utils,
      }
    }
    const searchQuery = 'a music search';
    const { input } = setup();
    fireEvent.change(input, { target: { value: searchQuery } })
    expect(input.value).toBe(searchQuery)
  });
});
