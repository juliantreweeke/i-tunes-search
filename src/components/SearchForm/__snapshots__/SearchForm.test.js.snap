// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm matches snapshot with loading prop 1`] = `
<div>
  <form
    class="searchbar"
  >
    <label
      for="searchQuery"
    >
      <span
        class="visually-hidden"
      >
        Enter search criteria here
      </span>
    </label>
    <div
      class="input_container"
    >
      <input
        class="input"
        id="searchQuery"
        name="searchQuery"
        placeholder="Search..."
        type="text"
        value=""
      />
    </div>
    <button
      class="button"
      data-loading="true"
      disabled=""
      type="submit"
    >
      <div
        class="loadingSpinnerContainer"
      >
        <div
          class="spinner"
          data-size="small"
          id="loading-spinner"
        />
      </div>
      Search
    </button>
  </form>
</div>
`;

exports[`SearchForm matches snapshot with placeholder prop 1`] = `
<div>
  <form
    class="searchbar"
  >
    <label
      for="searchQuery"
    >
      <span
        class="visually-hidden"
      >
        Enter search criteria here
      </span>
    </label>
    <div
      class="input_container"
    >
      <input
        class="input"
        id="searchQuery"
        name="searchQuery"
        placeholder="placeholder"
        type="text"
        value=""
      />
    </div>
    <button
      class="button"
      disabled=""
      type="submit"
    >
      Search
    </button>
  </form>
</div>
`;

exports[`SearchForm matches snapshot without placeholder prop 1`] = `
<div>
  <form
    class="searchbar"
  >
    <label
      for="searchQuery"
    >
      <span
        class="visually-hidden"
      >
        Enter search criteria here
      </span>
    </label>
    <div
      class="input_container"
    >
      <input
        class="input"
        id="searchQuery"
        name="searchQuery"
        placeholder="Search..."
        type="text"
        value=""
      />
    </div>
    <button
      class="button"
      disabled=""
      type="submit"
    >
      Search
    </button>
  </form>
</div>
`;
